File: caiyao.py
def output (data, t):
  dp = []*t

  # def dfs(index, left):
  #   key = "[%d,%d]"%(index,left)
  #   if (dp.has_key(key)):
  #     return dp[key]
  #   if (index >= len(data)):
  #     dp[key] = 0
  #     return 0
  #   dp[key] = max(
  #     dfs(index + 1, left),
  #     dfs(index + 1, left - data[index][0]) + data[index][1] if left >= data[index][0] else 0,
  #   )
  #   return dp[key]
    
  # return dfs(0 ,t)

  for d in data:
    dp_copy = dp.copy()
    for _t in range(t,d[1],-1):
      dp[_t] = max()
    for k,v in dp_copy.items():
      if k+d[0] <= t:
        dp[k+d[0]] = max(
          dp_copy[k+d[0]] if dp_copy.has_key(k+d[0]) else 0,
          v+d[1],
        )
        # print(d, n, dp)

  # print(dp)
  return max(dp.values())

t, n = map(lambda x:int(x), raw_input().split())
x = []
for i in range(0,n):
  x.append(map(lambda x:int(x), raw_input().split()))
# print(x)
print(output(x,t))

File: caiyao.js
const m = new Map;
function output(data, t) {
  function dfs(index, left) {
    const key = [index, left].join(',');
    if (m.has(key)) return m.get(key);
    if (index >= data.length) {
      m.set(key, 0);
      // m.set(key.join(','), 0);
      return 0;
    }
    const a = Math.max(
      dfs(index + 1, left),
      left > data[index][0] ? dfs(index + 1, left - data[index][0]) + data[index][1] : 0,
    )
    m.set(key, a);
    // m.set(key.join(','), a + node[1]);
    return a;
  }

  return dfs(0, t);
}

const r = output(
  // [
  //   [71, 100],
  //   [69, 1],
  //   [1, 2],
  // ],
  // 70
  [[71, 26], [34, 59], [82, 30], [23, 19], [1, 66], [88, 85], [12, 94]], 1000
);
console.log(r);
console.log([...m.entries()])